src/overflow/depression_fill.py

```diff
@@ -0,0 +1,255 @@
+from enum import Enum
+import numpy as np
+from numba import njit
+from osgeo import gdal
+from overflow.fill_pits import (
+    priority_flood_tile,
+    handle_edge,
+    handle_corner,
+    make_sides,
+)
+from overflow.setup_bands import setup_bands
+from overflow.util.raster import RasterChunk, raster_chunker, read_tile
+
+
+class Coordinate:
+    row: np.int32
+    col: np.int32
+
+    def __init__(self, row, col):
+        self.row = row
+        self.col = col
+
+
+class Side(Enum):
+    TOP = 1
+    RIGHT = 2
+    BOTTOM = 3
+    LEFT = 4
+
+
+class Corner(Enum):
+    TOP_LEFT = 1
+    TOP_RIGHT = 2
+    BOTTOM_RIGHT = 3
+    BOTTOM_LEFT = 4
+
+
+class DemLabelArr:
+    labels: np.ndarray
+    dem: np.ndarray
+
+
+class DemLabel:
+    label: int
+    dem: float
+
+
+@njit
+def depression_fill(input_path, output_path, chunk_size):
+    """
+    Fill depressions in a digital elevation model (DEM).
+    Algorithm 3 from Parallel Priority-Flood (R. Barnes, 2016)
+    """
+    with setup_bands(input_path, output_path) as bands:
+        input_band, output_band, no_data_value = bands
+        n_chunks_row, n_chunks_col = calculate_chunks(input_band, chunk_size)
+
+        for tile_index, tile_dem in enumerate(raster_chunker(input_band, chunk_size)):
+            sides = create_tile_side_flags(tile_dem, n_chunks_row, n_chunks_col)
+            tile_dem_labels, _ = priority_flood_tile(tile_dem, sides, no_data_value)
+            tile_name = get_tile_labels_file_name(tile_index)
+            tile_dem_labels.tofile(tile_name)
+
+        global_spillover_graph = connect_tile_edges_and_corners(
+            n_chunks_row, n_chunks_col, input_band, chunk_size
+        )
+        elevation_graph = priority_flood_graph(global_spillover_graph)
+
+        for tile_index, tile_dem in enumerate(raster_chunker(input_band, chunk_size)):
+            tile_name = get_tile_labels_file_name(tile_index)
+            tile_labels = np.fromfile(tile_name, dtype=int)
+            raise_elevation(output_band, tile_labels, tile_dem, elevation_graph)
+
+
+def connect_tile_edges_and_corners(n_tiles_row, n_tiles_col, input_band, chunk_size):
+    """Combine the spillover graphs of all tiles into a single global graph."""
+    spillover_graph = {}
+    tile_indexs_iter = walk_adjacent_tiles(n_tiles_row, n_tiles_col)
+    for top_l_i, top_r_i, bot_l_i, bot_r_i in tile_indexs_iter:
+        # + - - + - - +
+        # |  A  |  B  |
+        # + - - * - - +
+        # |  C  |  D  |
+        # + - - + - - +
+        tile_a = get_tile_labels_and_dem(top_l_i, n_tiles_col, input_band, chunk_size)
+        tile_b = get_tile_labels_and_dem(top_r_i, n_tiles_col, input_band, chunk_size)
+        tile_c = get_tile_labels_and_dem(bot_l_i, n_tiles_col, input_band, chunk_size)
+        tile_d = get_tile_labels_and_dem(bot_r_i, n_tiles_col, input_band, chunk_size)
+        # Connect A-B, A-C, D-B, D-C edges
+        combine_edge(tile_a, Side.RIGHT, tile_b, Side.LEFT, spillover_graph)
```

Comment:
Since we really only need the perimeter cells in this step, there is no need to read back the entire tile from disk. Can we just keep the perimeter cells for all tiles in memory after we run algo 1 over each tile to reduce IO?